metatensor-torch/src/module.cpp

```diff
@@ -0,0 +1,131 @@
+#include <torch/script.h>
+
+#include "metatensor/torch/labels.hpp"
+#include "metatensor/torch/block.hpp"
+#include "metatensor/torch/tensor.hpp"
+
+
+#include "metatensor/torch/module.hpp"
+
+
+using namespace metatensor_torch;
+
+template <typename T>
+bool is_custom_class(torch::IValue ivalue) {
+    if (ivalue.isCustomClass()) {
+        // this is inspired by the code inside `torch::IValue.toCustomClass<T>()`
+        auto* expected_type = c10::getCustomClassType<torch::intrusive_ptr<T>>().get();
+        return ivalue.type().get() == static_cast<const c10::Type*>(expected_type);
+    } else {
+        return false;
+    }
+}
+
+static std::pair<torch::IValue, bool> ivalue_to(
+    torch::IValue ivalue,
+    const torch::optional<at::Device>& device,
+    const torch::optional<at::ScalarType>& dtype,
+    bool non_blocking
+) {
+    if (is_custom_class<LabelsHolder>(ivalue)) {
+        if (device.has_value()) {
+            auto labels = ivalue.toCustomClass<LabelsHolder>();
+            labels = labels->to(device.value(), non_blocking);
+            return std::make_pair(labels, true);
+        } else {
+            return std::make_pair(ivalue, false);
+        }
+    } else if (is_custom_class<TensorBlockHolder>(ivalue)) {
+        auto block = ivalue.toCustomClass<TensorBlockHolder>();
+        block = block->to(dtype, device, non_blocking);
+        return std::make_pair(block, true);
+    } else if (is_custom_class<TensorMapHolder>(ivalue)) {
+        auto tensor = ivalue.toCustomClass<TensorMapHolder>();
+        tensor = tensor->to(dtype, device, non_blocking);
+        return std::make_pair(tensor, true);
+    } else if (ivalue.isGenericDict()) {
+        auto dict = ivalue.toGenericDict();
+        auto updated = c10::impl::GenericDict(dict.keyType(), dict.valueType());
+        auto all_changed = true;
+        auto some_changed = false;
+        for (const auto& item: dict) {
+            auto [updated_value, changed] = ivalue_to(item.value(), device, dtype, non_blocking);
+            all_changed &= changed;
+            some_changed |= changed;
+            updated.insert(item.key(), updated_value);
+        }
+        if (some_changed) {
+            if (!all_changed) {
+                C10_THROW_ERROR(ValueError,
+                    "dict containing both metatensor and non-metatensor data "
+                    "as values are not supported"
+                );
+            }
+            return std::make_pair(updated, true);
+        }
+    } else if (ivalue.isList()) {
+        const auto& list = ivalue.toList();
+        auto updated = c10::impl::GenericList(list.elementType());
+        auto all_changed = true;
+        auto some_changed = false;
```

Comment:
Minor nit again, I'd prefer these to be set to falsy and updated to truthy in the loop, i.e. `false and true` instead of true and false.